src/resmom/catch_child.c

```diff
@@ -296,28 +299,318 @@ static enum job_atr mom_rtn_list[] = {
+ *	specyfing a JSON object.
+ *
+ * @param[in]		value	-	string of JSON-object format
+ * @param[out]		msg 	-	error message buffer
+ * @param[in]		msg_len -	size of 'msg' buffer
+ *
+ * @return	PyObject *
+ * @retval	<python  object> -	dictionary representation of 'value'
+ * @retval	NULL		 -	if not successful, filling out 'msg'
+ *					with the actual error message.
+ */
+PyObject *
+json_loads(char *value, char *msg, size_t msg_len)
+{
+	PyObject	*py_name = (PyObject *)NULL;
+	PyObject	*py_module = (PyObject *)NULL;
+	PyObject 	*py_dict = (PyObject *)NULL;
+	PyObject	*py_func_loads = (PyObject *)NULL;
+	PyObject	*py_value = (PyObject *)NULL;
+	PyObject	*py_result = (PyObject *)NULL;
+
+	if (value == NULL) {
+		return ((PyObject *)NULL);
+	}
+
+	if (msg != NULL) {
+		if (msg_len <= 0) {
+			return ((PyObject *)NULL);
+		}
+		msg[0] = '\0';
+	}
+
+	py_name = PyString_FromString((char*)"json");
+	if (py_name == NULL) {
+		if (msg != NULL) {
+			snprintf(msg, msg_len, "failed to construct json name");
+		}
+		return ((PyObject *)NULL);
+	}
+
+	py_module = PyImport_Import(py_name);
+	if (py_module == NULL) {
+		if (msg != NULL) {
+			snprintf(msg, msg_len, "failed to import json");
+		}
+		goto json_loads_fail;
+	}
+
+	py_dict = PyModule_GetDict(py_module);
+	if (py_dict == NULL) {
+		if (msg != NULL) {
+			snprintf(msg, msg_len,
+				"failed to get json module dictionary");
+		}
+		goto json_loads_fail;
+	}
+
+	py_func_loads = PyDict_GetItemString(py_dict, (char*)"loads");
+	if ((py_func_loads == NULL) || !PyCallable_Check(py_func_loads)) {
+		if (msg != NULL) {
+			snprintf(msg, msg_len,
+				"did not find json.loads() function");
+		}
+		goto json_loads_fail;
+	}
+
+	py_value = Py_BuildValue("(z)", (char*)value);
+	if (py_value == NULL) {
+		if (msg != NULL) {
+			snprintf(msg, msg_len,
+				"failed to build python arg %s", value);
+		}
+		goto json_loads_fail;
+	}
+
+	PyErr_Clear(); /* clear any exception */
+	py_result = PyObject_CallObject(py_func_loads, py_value);
+
+	if (PyErr_Occurred()) {
+		if (msg != NULL) {
+			PyObject *exc_string = (PyObject *)NULL;
+			PyObject *exc_type = (PyObject *)NULL;
+			PyObject *exc_value = (PyObject *)NULL;
+			PyObject *exc_traceback = (PyObject *)NULL;
+
+			PyErr_Fetch(&exc_type, &exc_value, &exc_traceback);
+
+			/* get the exception */
+			if ((exc_type != NULL) &&
+				((exc_string = PyObject_Str(exc_type)) != NULL) &&
+				(PyString_Check(exc_string))) {
+				snprintf(msg, msg_len, "%s",
+					PyString_AsString(exc_string));
+			}
+			Py_XDECREF(exc_string);
+			Py_XDECREF(exc_type);
+			Py_XDECREF(exc_value);
+#if !defined(WIN32)
+			Py_XDECREF(exc_traceback);
+#elif !defined(_DEBUG)
+			/* for some reason this crashes on Windows Debug */
+			Py_XDECREF(exc_traceback);
+#endif
+		}
+		goto json_loads_fail;
+	} else if (!PyDict_Check(py_result)) {
+		if (msg != NULL) {
+			snprintf(msg, msg_len, "value is not a dictionary");
+		}
+		goto json_loads_fail;
+
+	}
+
+json_loads_pass:
```

Comment:
I understand your reasoning.  I had a similar thought about how it made the readability a little better.  My only worry is that labels aren't used much in C.  It might cause someone to go searching through the code to find a use of it.  Maybe turn it into a comment?